### PR TITLE
proximity_violation: added proximity violation check. also added assi…

### PR DESCRIPTION
**Pull request by p3ntak**
…gnedHospital attribute to the Resident class

### CODE REVIEW

**Comment by p3ntak** [comment]
for those looking to review this: it seemed like some of the requirements were not perfectly defined so I took a best guess at what we wanted the proximity checker to return and how we wanted it to behave. also please excuse the naiive null checks, i wasnt sure of the best way to do that in java. 

also we didnt have a place to store the assigned hospital for each resident and looking it up in the list of hospital assignments every time sounded expensive so i added an attribute for that.

File: src/main/java/edu/texas/social_computing/hospitals/Resident.java [hunk 1]
```diff
@@ -6,6 +6,15 @@
     private Resident partner;
     private List<Hospital> preferences;
     private List<Hospital> excludedHospitals;
+    private Hospital assignedHospital;
+
+    public Hospital getAssignedHospital() {
```
Comment (ericaddison) [inline src/main/java/edu/texas/social_computing/hospitals/Resident.java]: I thought we were going to have the assignment defined by the hospital's list of residents, and create a `getAssignmentForResident(List<Hospital>)` method in the utils class. That way we only track/change it in one place, less error prone.

File: src/main/java/edu/texas/social_computing/hospitals/Resident.java [hunk 2]
```diff
@@ -33,4 +42,24 @@ public Resident getPartner() {
...
+     */
+    public Resident worsePlacedPartner() {
+        Hospital hospital = this.assignedHospital;
+        Hospital partnerHospital = this.partner.getAssignedHospital();
+        if(hospital == null && partnerHospital == null) return this;
+        if(hospital == null && partnerHospital != null) return this.partner;
+        if(hospital != null && partnerHospital == null) return this;
+        int preference = this.preferences.indexOf(hospital);
+        int partnerPreference = this.partner.getPreferences().indexOf(partnerHospital);
+        return preference >= partnerPreference ? this : this.partner;
```
Comment (ericaddison) [inline src/main/java/edu/texas/social_computing/hospitals/Resident.java]: We might need to be careful with `>=` to make sure that it doesn't cause both partners to try and relocate if they get their same preference in different locations...

File: src/main/java/edu/texas/social_computing/hospitals/Resident.java [hunk 3]
```diff
@@ -33,4 +42,24 @@ public Resident getPartner() {
...
+    public boolean hasPartner() {
+        return this.partner != null;
+    }
+
+    /**
+     *
+     * @return the Resident that is matched this a Hospital that is lower on their preference list than their partner.
+     * The tie always goes to this
+     */
+    public Resident worsePlacedPartner() {
```
Comment (ericaddison) [inline src/main/java/edu/texas/social_computing/hospitals/Resident.java]: I would probably prefer to see this as a static method `worsePlacedResident(Resident r1, Resident r2)`

File: src/main/java/edu/texas/social_computing/hospitals/Resident.java (hunk 2 above)
Comment (ericaddison) [reply src/main/java/edu/texas/social_computing/hospitals/Resident.java]: Oh I see you took care of that case above :)

File: src/main/java/edu/texas/social_computing/hospitals/HRPP.java [hunk 4]
```diff
@@ -26,4 +33,36 @@
...
 
             // start from dominant partner's next hospital preference (changing the rank pointer)
+
+    /**
+     *
+     * @param residents: a list of residents in a matching that needs to be checked to see if any of the residents
+     *                 in the list violate a proximity constraint
+     * @return a set of residents that violate the proximity constraint AND are the worse placed partner
+     */
+    private Set<Resident> checkProximityViolations(List<Resident> residents) {
```
Comment (ericaddison) [inline src/main/java/edu/texas/social_computing/hospitals/HRPP.java]: Can we rename this to `getProximityViolations()` or `findProximityViolations()`, or something that better indicates the return value?

File: src/main/java/edu/texas/social_computing/hospitals/Resident.java (hunk 3 above)
Comment (p3ntak) [reply src/main/java/edu/texas/social_computing/hospitals/Resident.java]: you got it. that was the way i did it first. i changed it to this when i realized we would never need to do it for anyone that is not partners.

File: src/main/java/edu/texas/social_computing/hospitals/HRPP.java (hunk 4 above)
Comment (p3ntak) [reply src/main/java/edu/texas/social_computing/hospitals/HRPP.java]: yep. i was also thinking maybe `getProximityViolators()` but thought maybe that was too aggressive

Comment (ericaddison) [reply src/main/java/edu/texas/social_computing/hospitals/HRPP.java]: nice

File: src/main/java/edu/texas/social_computing/hospitals/Resident.java (hunk 1 above)
Comment (ericaddison) [reply src/main/java/edu/texas/social_computing/hospitals/Resident.java]: ok, then let's make static util methods `assign(Resident, Hospital)` and `unassign(Resident, Hospital)` that take care of both the Resident and Hospital fields at the same time.

Comment (ericaddison) [reply src/main/java/edu/texas/social_computing/hospitals/Resident.java]: Or manage the assignments externally, in a `Multimap` or something.

**Comment by p3ntak** [comment]
@ericaddison i updated this based on your comments and the most recent changes to master. lmk what you think (i probably did something wrong)

**Pull request closed without merging**